File: projet/client_leger/polyquiz/lib/services/game_config_service.dart
import 'package:flutter/material.dart';
import 'package:polyquiz/models/game_info_interface.dart';
import 'package:polyquiz/services/user_service.dart';
import 'package:polyquiz/models/user.dart';

class GameConfigService with ChangeNotifier {
  final UserService _userService = UserService();
  User? _currentUser;
  String? hostId;
  String gameType = '';
  int price = 0;
  bool friendsOnly = false;
  bool private = false;
  int prestige = 0;
  bool IA = false;

  GameConfigService([User? user]) {
    if (user != null) {
      _currentUser = user;
      hostId = user.uid;
    } else {
      hostId = 'null';
    }
  }

  void setUser(User user) {
    _currentUser = user;
    hostId = user.uid;
    notifyListeners();
  }

  void setGameType(String gameType) {
    this.gameType = gameType;
    notifyListeners();
  }

  void setPrice(int price) {
    this.price = price;
    notifyListeners();
  }

  void setPrestige(int prestige) {
    this.prestige = prestige;
    notifyListeners();
  }

  void setFriendsOnly(bool isFriends) {
    this.friendsOnly = isFriends;
    notifyListeners();
  }

  void setPrivacy(bool isPrivate) {
    this.private = isPrivate;
    notifyListeners();
  }

  void setIA(bool isAI){
    print('IA: $isAI');
    this.IA = isAI;
    notifyListeners();
  }

  GameConfig getGameConfig() {
    return GameConfig(
      hostUserId: hostId,
      gameType: gameType,
      private: private,
      onGoing: 'false', // this will change to 'true' when sent to server
      price: price,
      friendsOnly: friendsOnly,
      prestige: prestige,
      IA: IA,
    );
  }

  void reset() {
    gameType = '';
    price = 0;
    friendsOnly = false;
    private = false;
    notifyListeners();
  }
}

File: projet/client_lourd/src/app/services/game-config.service/game-config.service.ts
import { Injectable } from '@angular/core';
import {GameConfig} from "@common/interfaces/game-info.interface";
import {UsersService} from "@app/services/users.service/users.service";
import {Observable} from "rxjs";
import {User} from "@common/interfaces/user-data.interface";

@Injectable({
  providedIn: 'root'
})
export class GameConfigService {
  user$: Observable<User | null>;
  hostId: string | undefined;
  gameType: string;
  price: number;
  firendsOnly: boolean;
  private: boolean;
  prestige: number;


  constructor(private usersService: UsersService) {
    this.user$ = this.usersService.currentUserProfile$;
    this.user$.subscribe((user: User | null) => {
      this.hostId = user?.uid;
    });
  }

  setGameType(gameType: string) {
    this.gameType = gameType;
  }

  setPrice(price: number) {
    this.price = price;
  }

  setFriendsOnly(isFriends: boolean) {
    this.firendsOnly = isFriends;
  }

  setPrivacy(isPrivate: boolean) {
    this.private = isPrivate;
  }

  setPrestige(prestige: string) {
    this.prestige = Number(prestige);
  }

  getGameConfig() {
    return {
      hostUserId: this.hostId,
      gameType: this.gameType,
      private: this.private,
      onGoing: false, // this will change to true when sent in server
      price: this.price,
      friendsOnly: this.firendsOnly,
      prestige: this.prestige
    } as unknown as GameConfig;
  }

  reset() {
    this.gameType = '';
    this.price = 0;
    this.firendsOnly = false;
  }
}
